Add first version of new reduced lung module

bgoderbauer opened the pull request:
<!--
* Title: Provide a general summary of your changes in the Title above.

* Assignees:  If you know anyone who should likely handle bringing this pull request (PR) to completion -- that's usually yourself -- select them from the Assignees drop-down on the right.

* Labels: Update the label of the issue(s) addressed by this pull request to "Under Review".
-->

## Description and Context
<!--
Provide a brief and concise description of your proposed change. Questions you should think about:
* Why is this change required?  What problem does it solve?
* Is there a bigger picture? Is this PR a part of a larger set of changes? Which further steps are planned after merging this PR, if any?
* How has the proposed implementation been verified and tested?

Keep the description of the PR always up-to-date and concise.
-->
This MR adds a new code basis for the 0D lung model equations found in `red_airways` with the central goal of being more performant and easier to maintain and extend than the current version. For now, this implementation can only calculate lung models consisting of resistive airways (Hagen-Poiseuille flow) and terminal units with linear Kelvin-Voigt viscoelasticity. The input file and parameter description are also built upon the existing version, so both versions of the code can be executed by interchanging `Reduced_Lung` and `ReducedDimensionalAirWays` for the problem type.

Key differences compared to the old version:

- The system of equations is solved via a Newton-Raphson iteration, so, depending on the underlying models, a lot less nonlinear iterations until convergence should be necessary compared to the fixed-point iteration.
- All junction equations connecting the reduced-order elements are explicitly calculated to allow for more transport models over the element boundaries (similar to DG). Also, the flow inside each reduced-order element is treated as an additional degree of freedom. Overall, the system of equations is ~3 times larger than the old version.
- Evaluating "element" equations no longer uses the 4C element infrastructure. 

This first version is still highly WiP, but it works as intended and shall serve as a starting point for future developments and major reworks/extensions.

A direct comparison between the old version and the new implementation on a rather large lung example with ~100'000 reduced-order elements already shows a speedup of 20-30x per iteration step.

bgoderbauer commented on src/reduced_lung/4C_reduced_lung_helpers.cpp:
```diff
@@ -0,0 +1,177 @@
+
+FOUR_C_NAMESPACE_OPEN
+
+Epetra_Map create_locally_owned_dof_map(const Epetra_Comm& comm, const std::vector<Airway>& airways,
```
Yes, the names of the maps are currently different than the usual names in 4C. I'm open to changes here, but I wanted to clarify the map's purpose since it took a while for me to really understand what they should do. 
The significant difference compared to most other map layouts in 4C lies in the "locally owned dof map". It is independent of the row map of the system matrix and instead a "domain map". That's also why the name "matrix row map" indicates that the row-wise layout only accounts for the system matrix and right-hand-side vector. 
Yes, the "locally relevant dof map" is a column map.

mayrmt replied:
@bgoderbauer I am surprised, that the maps are different on purpose. Why is that? Due to the overlapping domain decomposition in 4C, I'd expect that row map == range map == domain map in almost all cases. Especially, the vector of unknowns usually is based on the row layout of the matrix.

bgoderbauer replied:
The problem is that, unfortunately, domain and system equations aren't directly tied together and aren't as nice to handle as most FE formulations. As an example, each reduced-order "element" introduces three new degrees of freedom but only one state equation. The remaining equations are couplings or boundary conditions. 
I want to briefly explain my intention and thought process behind this structuring:
The total dofs are my starting point for domain decomposition as it makes sense to group locally close and connected elements as much as possible. However, the assignment of coupling equations would be, at some point, arbitrary if I use the same map for the matrix rows as my domains (sometimes they start from a child airway, sometimes from a parent airway, and sometimes they may even be assigned to a rank where none of the dofs is present). The graph-like structure of airway trees is somewhat arbitrary. That's why I've decided (with a lot of help from @sebproell) to follow a custom logic when assigning the system equations to the row map that can be checked and interpreted more easily. In theory, this should also lead to minimal overlap in the column map that needs to be communicated. 
IMO it would be possible to use the same maps with advanced logic for distributing the dofs and/or equations, but I think it is a lot harder to achieve this in a consistent manner while keeping communication low and being able to understand the assignment process. Maybe an interesting optimization problem for a student project.

bgoderbauer replied:
So TLDR: A rank owing n dofs may not have n equations following the implemented logic

mayrmt replied:
Interesting! Thanks for the information.

bgoderbauer commented on src/reduced_lung/4C_reduced_lung_main.cpp:
```diff
@@ -0,0 +1,764 @@
+    double err_norm = tolerance + 1;  // Error norm for convergence.
+    int i = 1;                        // Counter for nonlinear iterations.
+    // Nonlinear loop.
+    while (err_norm >= tolerance && i <= max_nonlinear_iterations)
```
Thanks for the hint; I didn't think of that yet! I want to reuse general infrastructure wherever it is suitable, so I'll definitely have a look at this in the future. In the current version, the time and nonlinear loop are very basic; there is a lot of work left to do.

mayrmt replied:
Sure, I understand. As a matter of fact, a simple Newton loop is always easy to write, but it still needs to be maintained. In order to have a Newton iteration with all nice features (inexact Newton methods, line search, detailed convergence checks, maybe even other nonlinear solvers than just plain Newton-Raphson) is way harder, so we never do it ourselves. If you use a suitable package from the start, you have access to all these features. If one goes with a self-coded Newton, experience shows that one sticks with it for eternity.

sebproell commented on src/reduced_lung/4C_reduced_lung_main.cpp:
```diff
@@ -0,0 +1,764 @@
+        std::array<double, 3> vals;
+        double res;
+        evaluate_aw_jacobian(vals, airway.local_airway_id);
+        evaluate_aw_rhs(res, locally_relevant_dofs[airway.local_dof_ids[0]],
+            locally_relevant_dofs[airway.local_dof_ids[1]],
+            locally_relevant_dofs[airway.local_dof_ids[2]], airway.local_airway_id);
```
You could name the magic numbers 0,1 and 2 here, e.g.

```
enum DofName{
  flow = 0,
  pressure_in  = 1,
  pressure_out = 2,
};
```

sebproell commented on src/reduced_lung/4C_reduced_lung_main.cpp:
```diff
@@ -0,0 +1,764 @@
+  for (const auto* ele : actdis->my_row_element_range())
+  {
+    const auto* nodes = ele->nodes();
+    // WARNING: if node ordering is wrong (inlet in nodes[1]), the whole process doesn't work! Needs
+    // to be enforced during tree creation.
```
Could you assert that this assumption holds?

mayrmt commented on src/reduced_lung/4C_reduced_lung_helpers.hpp:
```diff
@@ -0,0 +1,195 @@
+    const std::vector<BoundaryCondition>& boundary_conditions);
+
+/*!
+ * @brief Create the Epetra map with the dof indices relevant for the locally owned SoE rows.
```
I am wondering: Shouldn't "SoE" be "EoS" (=Equation of State)? What does "SoE" stand for?

bgoderbauer replied:
I changed it. It should have been "system of equations," but this is probably more of an uncommon abbreviation that I often use.

bgoderbauer commented:
I've added most of the suggestions - thanks, @mayrmt and @sebproell! I would opt to merge the status as it is to have a basis where others and I can implement the necessary extensions and improvements.

Regarding the Petra objects: I find the Trilinos naming of domain, row, and column map quite intuitive for an abstract description of their purpose, and I understand their high recognition value when trying to understand different parts of 4C. So, I changed the naming of the functions that create these maps according to this standard. In the implementation, however, I find it more intuitive to see their purpose directly regarding the assembly of the system of equations, i.e., distributing the ownership of dofs and the relevance of dofs across ranks. I hope this is understandable!

The next steps will be:

1. Use a standard nonlinear solver (probably NOX) and prepare the necessary steps.
2. Design suitable abstractions for the different models (e.g., resistive vs. compliant airways).
3. Create VTK output for this problem

mayrmt commented on src/reduced_lung/4C_reduced_lung_helpers.hpp:
```diff
@@ -0,0 +1,199 @@
+    int local_connection_id;
+    int global_parent_element_id;
+    int global_child_element_id;
+    // dofs: {p2_parent, p1_child, q2_parent (q for non-compliant airways), q1_child}
+    std::array<int, 4> global_dof_ids;
+    std::array<int, 4> local_dof_ids{};
```
Can this encoding of the indices of these arrays be given more explicitly, e.g. though an `enum`? Right now, we solely depend on a comment and always have to look up the meaning of each index.

(The same comment applies to other `struct`s in this file.)

bgoderbauer replied:
I've added the DofNumbering enum inside the respective structs (I didn't know this was possible before :D). Now, it should be easier to read. Some things need further abstractions in the future (e.g. the order of equations inside a bifurcation), but I think this will develop naturally with the next steps.

mayrmt commented on src/reduced_lung/4C_reduced_lung_main.cpp:
```diff
@@ -0,0 +1,773 @@
+      // the coupling entities. However, special cases might slip through.
+      const auto& node_in = *nodes[0];
+      const auto& node_out = *nodes[1];
+      // const auto node_in_n_eles = global_ele_count_per_node[node_in.id()];
+      const auto node_in_n_eles = global_ele_ids_per_node[node_in.id()].size();
+      // const auto node_out_n_eles = global_ele_count_per_node[node_out.id()];
+      const auto node_out_n_eles = global_ele_ids_per_node[node_out.id()].size();
```
```suggestion
      const auto node_in_n_eles = global_ele_ids_per_node[node_in.id()].size();
      const auto node_out_n_eles = global_ele_ids_per_node[node_out.id()].size();
```

mayrmt commented on src/reduced_lung/4C_reduced_lung_main.cpp:
```diff
@@ -0,0 +1,773 @@
+    std::vector<double> area(n_airways);
+    std::vector<double> poiseuille_resistance(n_airways);
+
+    // Airway constants. Read from input file?
```
??

mayrmt reviewed (changes requested):
@bgoderbauer I have left some ideas for improvement, some small request for changes as well as some questions, where I just don't understand yet, why certain choices have been made. Overall, this looks pretty good.

bgoderbauer commented on src/reduced_lung/4C_reduced_lung_main.cpp:
```diff
@@ -0,0 +1,773 @@
+                << "----------------------------------------\n"
+                << std::flush;
+    }
+    for (int n = 1; n <= n_timesteps; n++)
```
I haven't considered the right abstraction for time loops, but I'd be glad to hear ideas/suggestions here! In the future, there shall be functionality for adaptive time stepping with advanced models (Carolin Geitners last extensions regarding alveolar recruitment/derecruitment very likely need a flexible time stepping mechanism. Her model will definitely be included in the new implementation).

bgoderbauer commented on src/reduced_lung/4C_reduced_lung_main.cpp:
```diff
@@ -0,0 +1,773 @@
+
+        // Currently linear system of equations. Assembly still highly WIP. Error update will depend
+        // on rhs when assembly process is clearly separated.
+        err_norm = 1e-10;
```
I removed the nonlinear pseudo loop. It helped me group the different steps during implementation, but it is unnecessary at this stage and may be confusing.

bgoderbauer commented on src/reduced_lung/4C_reduced_lung_helpers.hpp:
```diff
@@ -0,0 +1,211 @@
+    int local_equation_id;
+    int local_airway_id;
+    AirwayType airway_type;
+    // dofs: {p1, p2, q} for resistive airways; {p1, p2, q1, q2} for compliant airways
+    std::vector<int> global_dof_ids{};
```
The number of dofs can vary depending on the model. A compliant airway wall, for example, introduces 1-2 new dofs as the volume change directly influences flow inside the airway. I created a little helper with the `ElementDofNumbering` enum, but we haven't finalized the abstraction for different models.

bgoderbauer commented on src/reduced_lung/4C_reduced_lung_helpers.hpp:
```diff
@@ -0,0 +1,211 @@
+    int global_parent_element_id;
+    int global_child_1_element_id;
+    int global_child_2_element_id;
+    enum DofNumbering
+    {
+      p_out_parent = 0,
+      p_in_child_1 = 1,
+      p_in_child_2 = 2,
+      q_out_parent = 3,
+      q_in_child_1 = 4,
+      q_in_child_2 = 5
+    };
+    std::array<int, 6> global_dof_ids;
+    std::array<int, 6> local_dof_ids{};
```
I think it's important here since these enums specify the dof ordering inside the element. This way, I don't need to know this ordering when assembling the matrix.
